feat: Add trigger to combobox #1136

aalencar opened the pull request:
* Add prop trigger to combobox allowing the choice to be sent to the server right away when it is selected.

resolves #1136

aalencar commented on py/examples/combobox.py:
```diff
@@ -24,6 +25,8 @@ async def serve(q: Q):
                         choices=combobox_choices, disabled=True),
             ui.combobox(name='combobox_error', label='Enter or choose a color', placeholder='Color...', value='Blue',
                         choices=combobox_choices, error='This combobox has an error!'),
+            ui.combobox(name='combobox_trigger', trigger=True, label='Enter or choose a color (trigger on selection)', placeholder='Color...', value=q.args.combobox_trigger,
```
@mturoci is there something wrong with using the value that was recently send to the server? And how do I conditionally pass arguments in Python like in JS `value || fallback`?

lo5 replied:
It's confusing to read. It's enough if the example line demonstrates how to make a combobox submit its value when changed. The `value=q.args.combobox_trigger` is unnecessary.

mturoci replied:
>is there something wrong with using the value that was recently send to the server

`value` is meant to be "initial" value - all wave components should be uncontrolled. That means the `value` would work only if you recreate the card.

>how do I conditionally pass arguments in Python 

`value or fallback`

mturoci replied:
Anyway it's better to create a separate example for `trigger`, like we have for `textbox` for example. The reason is that triggering will submit also filled values from other comboboxes.

aalencar replied:
I added a new example like we have for textbox

lo5 commented on py/examples/picker.py:
```diff
@@ -14,7 +14,7 @@ async def serve(q: Q):
     else:
         q.page['example'] = ui.form_card(box='1 1 4 10', items=[
-            ui.picker(name='picker', label='Place an order (try Spam, Eggs or Ham):', choices=[
+            ui.picker(name='picker', trigger=True, label='Place an order (try Spam, Eggs or Ham):', choices=[
```
This example is unrelated to the combobox. Is this an intentional change?

aalencar replied:
I was checking the picker behavior on trigger and commited it by mistake. Reverted, thanks.

lo5 commented on py/examples/combobox.py:
```diff
@@ -24,6 +25,8 @@ async def serve(q: Q):
                         choices=combobox_choices, disabled=True),
             ui.combobox(name='combobox_error', label='Enter or choose a color', placeholder='Color...', value='Blue',
                         choices=combobox_choices, error='This combobox has an error!'),
+            ui.combobox(name='combobox_trigger', trigger=True, label='Enter or choose a color (trigger on selection)', placeholder='Color...', value=q.args.combobox_trigger,
```
```suggestion
            ui.combobox(name='combobox_trigger', trigger=True, label='Enter or choose a color (submit color when changed)', placeholder='Color...', value='Blue',
```

aalencar commented:
![2021-12-09 12 49 00](https://user-images.githubusercontent.com/15820796/145391466-510c0507-c5a6-4fb1-b644-8ab02ec91c96.gif)

mturoci commented on py/examples/combobox.py:
```diff
@@ -14,6 +14,7 @@ async def serve(q: Q):
             ui.text(f'combobox={q.args.combobox}'),
             ui.text(f'combobox_disabled={q.args.combobox_disabled}'),
             ui.text(f'combobox_error={q.args.combobox_error}'),
+            ui.text(f'combobox_trigger={q.args.combobox_trigger}'),
```
out

aalencar replied:
Removed.

mturoci commented on ui/src/combobox.tsx:
```diff
@@ -51,6 +51,8 @@ export interface Combobox {
   visible?: B
   /** An optional tooltip message displayed when a user clicks the help icon to the right of the component. */
   tooltip?: S
+  /** True if the choice should be submitted when the dropdown value changes. */
```
Not just dropdown.

aalencar replied:
Improved.

mturoci reviewed (changes requested):
Thanks @aalencar. Don't forget that combobox also allows free text (not just dropdown) which needs to be triggered as well. The free text needs to be debounced so that we do not submit every keystroke. See `textbox` for an implementation reference.

aalencar commented:
> The free text needs to be debounced so that we do not submit every keystroke

I implemented the debounce but I noticed that `onChange` is only triggered on blur, pressing enter,  or selecting a choice.
Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.

mturoci commented:
>I implemented the debounce but I noticed that onChange is only triggered on blur, pressing enter, or selecting a choice.

Good to know.

>Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.

I don't understand this one. The value will never be undefined thanks to `const v = option?.text || value || ''`, right?

aalencar commented:
> > I implemented the debounce but I noticed that onChange is only triggered on blur, pressing enter, or selecting a choice.
> 
> Good to know.
> 
> > Based on that, there is no need to debounce if we prevent the user from sending undefined choices. I added a guard to prevent that.
> 
> I don't understand this one. The value will never be undefined thanks to `const v = option?.text || value || ''`, right?

By undefined I mean some choice that is not present in choices / dropdown (i.e. not defined by the user).

Example: `choices = ['red', 'yellow', 'green']`, if user types `blue` or `yello` it won't send it to the server

 I made it more confusing by using the word 'undefined' in a javascript context. I hope it's more clear now.

mturoci commented:
But this is not the correct logic, combobox allows free text as well which should be sent to server also. Otherwise, there would be no difference between combobox and dropdown basically. Please revert the condition.

aalencar commented:
> But this is not the correct logic, combobox allows free text as well which should be sent to server also. Otherwise, there would be no difference between combobox and dropdown basically. Please revert the condition.

Reverted.

mturoci reviewed (approved):
Thanks @aalencar!
